### DISCUSSION TITLE
Vídeo 02 - Configuração

### DISCUSSION THREAD

**Discussion by thpoiani** [thpoiani/lab-quarkus#2] [General]
Olá novamente! Nesta aula trabalhamos com contêineres e preparamos o ambiente para deployment.
Desenvolvemos um Pipeline de _Continuous Integration and Continuous Deployment_ rústico, mas que será útil para os conceitos da Aula.

Agora tenho uma pergunta para você: **Quais são possíveis melhorias para esse processo de CICD?**

Pense fora da caixa e sugira qualquer coisa.

<details>
  <summary>Exemplos (abra depois de comentar para prevenir viés):</summary>

- "Use uma plataforma de CICD, como [Jenkins](https://www.jenkins.io/), [Github Actions](https://github.com/features/actions) ou até mesmo [GitLab CI/CD](https://docs.gitlab.com/ee/ci/)"
- "Considere outra estratégia de _progressive deployment_. **Canary** é uma excelente alternativa"
</details>

🌟 Aproveitando o assunto, sugiro fortemente que acesse o documento de **[Deployment Pipeline Reference Architecture](https://pipelines.devops.aws.dev/)** da Amazon

**Comment by thpoiani** [comment]
Você conhece [maven-release-plugin](https://maven.apache.org/maven-release/maven-release-plugin/index.html)? Esse plugin pode simplificar o processo de build e release para projetos maven.

**Reply by didifive** [reply]
Ainda não utilizei este recurso, já deixando na lista para entender e praticar

**Reply by ale-fuentes-ar** [reply]
Boa tarde estimado @thpoiani .. eu nao utilicei ele ainda ...
Desculpe, poderia indicar algum tutorial ?

**Reply by thpoiani** [reply]
Olá @ale-fuentes-ar, sugiro que você leia a própria documentação do maven-release-plugin: https://maven.apache.org/maven-release/maven-release-plugin/usage.html 👍

**Comment by didifive** [comment]
Pensando sobre o script atual, poderia ter mais comando de git como um commit que já adicionava a atualização de outros arquivos do projeto e que já fizesse um push para o repositório remoto para deixá-lo atualizado.
Uma coisa que adicionei até o momento no cicd-build.sh, foi para que quando fosse gerada a imagem da api para o docker, utilizando o sed do linux, já altera a tag do docker compose para quando for subir as aplicações por ele já subir a última versão do build.
Fora isso seria utilizar ferramentas como o GitHub Actions ou outras automações para CI/CD

**Comment by mauromortoni** [comment]
. Verificação de Argumentos
. Uso de Funções
. Tratamento de Erros
. Variáveis Descritivas
. Comentários

Script (não foi testado)

#!/bin/bash -xe

# Função para exibir mensagem de erro e sair
function error_exit {
    echo "$1" 1>&2
    exit 1
}

# Verifica se o nome do aplicativo foi passado como argumento
if [ -z "$1" ]; then
    error_exit "Uso: $0 <nome-do-aplicativo>"
fi

APP=$1
ROOT=$(pwd)

# Verifica se o diretório do aplicativo existe
if [ ! -d "$APP" ]; then
    error_exit "O diretório $APP não existe."
fi

# Muda para o diretório do aplicativo
cd "$APP" || error_exit "Falha ao entrar no diretório $APP"

# Verifica se o mvnw existe e é executável
if [ ! -x "./mvnw" ]; then
    error_exit "O arquivo mvnw não existe ou não é executável."
fi

# Limpa e prepara o projeto
./mvnw clean || error_exit "Falha ao executar mvnw clean"
./mvnw versions:set -DremoveSnapshot || error_exit "Falha ao remover snapshot"
APP_VERSION=$(./mvnw -q -Dexec.executable=echo -Dexec.args='${project.version}' --non-recursive exec:exec) || error_exit "Falha ao obter versão do aplicativo"

# Valida se a versão foi obtida corretamente
if [ -z "$APP_VERSION" ]; then
    error_exit "Não foi possível obter a versão do aplicativo."
fi

./mvnw package || error_exit "Falha ao empacotar o aplicativo"
./mvnw versions:set -DnextSnapshot || error_exit "Falha ao definir próximo snapshot"

# Verifica se há alterações não comitadas no Git
if [[ -n $(git status --porcelain) ]]; then
    git add pom.xml || error_exit "Falha ao adicionar pom.xml ao Git"
    git commit -m "cicd: bump version ${APP}:${APP_VERSION}" || error_exit "Falha ao comitar alterações no Git"
else
    echo "Nenhuma alteração para commit no Git."
fi

# Retorna ao diretório raiz
cd "$ROOT" || error_exit "Falha ao retornar ao diretório raiz"

# Verifica se o Docker está instalado e em execução
if ! command -v docker &> /dev/null; then
    error_exit "O Docker não está instalado."
fi
if ! docker info &> /dev/null; then
    error_exit "O Docker não está em execução."
fi

# Constrói a imagem Docker
TAG=$APP_VERSION docker-compose build --no-cache "$APP" || error_exit "Falha ao construir a imagem Docker"

# Lista as imagens Docker criadas
docker images "dio/${APP}" || error_exit "Falha ao listar imagens Docker"
